### ISSUE TITLE
New lint: Unnecessarily complex types

### ISSUE THREAD

**Issue by jyn514** [rust-lang/rust-clippy#4844]
It would be nice to lint on return types that are more complex than necessary. Take for example the following function ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=446cb1042ea60148198aaa1100e35d6b):

```rust
fn unnecessary_result() -> Result<usize, ()> {
    Ok(1)
}
```

The `Result` here is not necessary, because the function always returns an `Ok` variant.
I would also like to see a similar lint for `fn f() -> Option<T>` where `f` always returns `Some`.

Note this is _not_ the same as https://rust-lang.github.io/rust-clippy/current/index.html#type_complexity, which is about return types that are hard to read, not return types that are unnecessary.

Possible output for the lint (just making this up here):

```rust
warning: unnecessary `Result` in return type
 --> src/lib.rs:1:1
   |
1 |  fn unnecessary_result() -> Result<usize, ()> {
   |                                             ^^^^^^^^^^^^^^^^ help: try removing the `Result` call: `usize`
2 |         Ok(1)
   |         ^^^^^ note: only return value is an `Ok` variant
```

**Comment by jfrimmel** [comment]
Sometimes such a construct is necessary, e.g. to stay SemVer-compatible (e.g. you changed the internals of a function, so that it now cannot fail anymore, but your users still expect a `Result`/`Option`).
Another possibility is, that you want to be future-proof: currently the implementation cannot fail, but it is expected to be so in the future.
I still think, that this might be a useful lint, but the SemVer hazard should be documented in "known problems".

**Comment by jyn514** [comment]
Maybe it could be 'allow' by default and you'd have to explicitly set it to `warn`.

**Comment by flip1995** [comment]
Definitely an `allow-by-default` lint. It may be hard to determine all the return points in the function and find out what exactly is returned. Making this generic over multiple types may also be quite hard.

**Comment by mikerite** [comment]
`unnecessary_filter_map` already does something similar. It warns it the closure never returns `None`.
